libsmi/tools/smidump.1
.\"
.\" $Id: smidump.1.in 7608 2008-01-22 13:05:41Z schoenw $
.\"
.TH smidump 1  "August 10, 2004" "IBR" "SMI Tools"
.SH NAME
smidump \- dump SMI or SPPI modules in various formats
.SH SYNOPSIS
.B smidump
[
.B "-Vhqusmk"
] [
.BI "-c " file
] [
.BI "-o " name
] [
.BI "-p " module
] [
.BI "-l " level
] [
.BI "-f " format
]
.I "module(s)"
.SH DESCRIPTION
The \fBsmidump\fP program is used to dump the contents of a single MIB
or PIB
module or a collection of modules to stdout in a selectable output
format.  This format may be a simple tree of nodes, types or imported
modules, but also a format fully compliant to SMIv1, SMIv2, SPPI or SMIng or
CORBA IDL or C source code. Smidump can thus be used to convert
modules from SMIv2 to SMIng and from SMIng to SMIv2, or to develop
template based agent code.
.SH OPTIONS
.TP
\fB-V, --version\fP
Show the smidump version and exit.
.TP
\fB-h, --help\fP
Show a help text and exit. The help text contains a list of all
supported output formats.
.TP
\fB-c \fIfile\fB, --config=\fIfile\fP
Read \fIfile\fP instead of any other (global and user)
configuration file.
.TP
\fB-f \fIformat\fB, --format=\fIformat\fP
Use \fIformat\fP when dumping a module. Supported output formats are described 
below. The default output format is SMIng. The \fIformat\fP argument is
case insensitive.
.TP
\fB-l \fIlevel\fB, --level=\fIlevel\fP
Report errors and warnings up to the given severity \fIlevel\fP. See
the smilint(1) manual page for a description of the error levels. The
default error level is 3.
.TP
\fB-s, --severity\fP
Show the error severity in brackets before error messages.
.TP
\fB-m, --error-names\fP
Show the error names in braces before error messages.
.TP
\fB-o \fIname\fB, --output=\fIname\fB
Write the output in one or multiple files instead of stdout. The file
name(s) are derived from the \fIname\fP argument.  Not all format
support this option.
.TP
\fB-p \fImodule\fB, --preload=\fImodule\fP
Preload the module \fImodule\fP before reading the main module(s). This may
be helpful if an incomplete main module misses to import some definitions.
.TP
\fB-q, --quiet\fP
Suppress comments from dumped modules. What kind of information
gets suppressed depends on the output format.
.TP
\fB-u, --unified\fP
Dump a unified output in case of multiple \fImodule(s)\fP instead of
multiple concatenated output sections. This is not supported for all
output formats.
.TP
\fB-k, --keep-going\fP
Continue as much as possible after serious parse errors. Note that
the output generated after serious parse errors may be incomplete
and should be used with care.
.TP
.I module(s)
These are the module(s) to be dumped. If a module argument represents a
path name (identified by containing at least one dot or slash character),
this is assumed to be the exact file to read. Otherwise, if a module is
identified by its plain module name, it is searched according to libsmi
internal rules. See \fBsmi_config(3)\fP for more details.
.SH "OUTPUT FORMATS"
The \fBsmidump\fP program supports the following output formats:
.TP 12
sming
SMIng compliant format as defined in the SMIng Internet-Draft.
.TP
smiv2
SMIv2 compliant format as defined in RFC 2578, RFC 2579, RFC 2580.
.TP
smiv1
SMIv2 compliant format as defined in RFC 1155, RFC 1212, RFC 1215.
.TP
sppi
SPPI compliant format as defined in RFC 3159. If the underlying module
is not SPPI, there might be some mandatory information missing.
.TP
mosy
Format generated by the mosy compiler.
.TP
imports
Import hierarchy of a module.
.TP
types
Types defined in a module.
.TP
tree
OID registration tree structure of a module.
.TP
metrics
Metrics derived from a module (experimental).
.TP
identifiers
List of identifiers defined in a module.
.TP
compliances
Compliance definitions with all included objects and notifications.
.TP
corba
CORBA IDL and OID definitions following the JIDM specification translation
rules.
.TP
netsnmp
C source code files for usage within the net-snmp package (experimental).
.TP
scli
ANSI C manager stubs for usage within the scli package (experimental).
.TP
cm
Reverse engineered conceptual model in DIA XML file format (experimental).
.TP
svg
SVG diagram of a module (experimental). Use with \fB-u\fP when dumping multiple \fImodules\fP.
.TP
jax
Java AgentX sub-agent classes in separate files (experimental).
.TP
perl
Perl represention of the MIB module
(contributed by Martin Schulz <schulz@videotron.ca>).
.TP
python
Python dictionaries represention of the MIB module
(contributed by Pat Knight <pat@ktgroup.co.uk>).
.TP
xml
SMI in XML format (experimental).
.TP
xsd
SMI in XML schema format (experimental).
.TP
sizes
SNMP best case / worst case RFC 3416 PDU sizes for typical PDUs
excluding SNMP message and transport headers (experimental).

.SH "EXAMPLE"
This example converts the SMIv2 module IF-MIB in the current directory
to IF-MIB.sming in SMIng format. Note that the ./ prefix is used to
ensure reading the module from the current directory and not from a
place that libsmi guesses on its own.
.nf

  $ smidump -f sming ./IF-MIB > IF-MIB.sming
.fi
.SH "SEE ALSO"
The
.BR libsmi (3)
project is documented at
.BR "http://www.ibr.cs.tu-bs.de/projects/libsmi/" "."
.PP
.BR smilint "(1)"
.SH "AUTHORS"
(C) 1999-2004 F. Strauss, TU Braunschweig, Germany <strauss@ibr.cs.tu-bs.de>
.br
(C) 1999-2002 J. Schoenwaelder, TU Braunschweig, Germany <schoenw@ibr.cs.tu-bs.de>
.br
(C) 2002-2003 J. Schoenwaelder, University of Osnabrueck, Germany
.br
(C) 2003-2004 J. Schoenwaelder, International University Bremen, Germany
.br
(C) 2001-2002 T. Klie, TU Braunschweig, Germany <tklie@ibr.cs.tu-bs.de>
.br
(C) 2002 M. Bunkus, TU Braunschweig, Germany <bunkus@ibr.cs.tu-bs.de>
.br
and contributions by many other people.
.br


qwt/CHANGES
Release 5.1.0
===================

Changes
-------
1) QwtSymbol::copy introduced
   Now it is possible to use derived symbol classes for curves
2) QwtPlotScaleItem introduced
   A new type of plot item for displaying axes on the canvas
3) QwtClipper added
   A collection of clipping algos
4) Using DBL_EPSILON
   This change allows smaller intervals for sliders/dials
5) QwtPanner
   setOrientation() added.
6) QwtPlot
   axisStepSize() added
   clear is virtual now
7) QwtPlotPrintFilter
   PrintCanvasBackground splitted into PrintBackground, PrintFrameWithScales
8) QwtPlotZoomer
   setZoomStack() added
9) Changes for the QwtPolar package
   QwtLegendItemManager introduced
   QwtMagnifier introduced
10)Suffix rules added in qwtconfig.pri for different targets for 
   debug/release builds.

Bug Fixes
---------
1. QwtAbstractScaleDraw::setAbstractScaleDraw
   Reinitialization problem fixed
2. QwtLegendItem
   key event handlers fixed
3. QwtPicker
   solaris-cc compiler problem fixed
4. Inaccurate mapping of scale to widget coordinates fixed
5. QwtPlotCurve::draw
   Updates for Qt 4.3 added
6. QwtPlotLayout
   AlignToCanvas layout calculation fixed
7. QwtPlot::print
   Workaround for a QPen initialization problem, 
   when printing to Pdf, added
8. QwtText
   Layout of rich text documents fixed
9. Designer
   Handling of QwtScaleWidget fixed
10. realtime example
   Qt::WA_PaintOutsidePaintEvent added, ScrollZoomer fixed
11. Several others I have forgotten

Release 5.0.2
===================

Bug Fixes
---------
1. QwtPlotCurve::Xfy curve type fixed
2. Memory leak in QwtLegend fixed
3. Vertical alignment of rich texts fixed
4. Workaround for a Qt4 bug added, that produces horrible performance
   when painting curves with a pen width > 1.
5. Background for the tracker text of QwtPickers fixed.
   Improved (faster + better rendered texts) implementation of
   painting tracker texts, using capabilities of Qt >= 4.3.
6. QwtArrowButton/QwtCounter: workaround for layout bug ( Qt < 4.3 )
   of the Cleanlook style added.
7. A couple of minor fixes

Changes
-------
1. QSvgGenerator added to the bode example

Release 5.0.1
===================

Changes
-------
1.  A couple of problems, when building Qwt fixed.
2.  Displaying Rich Text with Qt 4.x fixed

Release 5.0.0
===================

Platforms
---------
Support of Qt3 and Qt4. Qt2 is not supported any longer.

Key features
------------
1.  Redesign of plot items. Makes it much easier to develop
    individual items.
2.  Redesign of the scale classes. All calculations are
    collected in scale engines, where the application can
    implement it´s own (f.e log2, or date scales). Now it´s
    also possible to have individual and completely irregular scales
3.  Redesign of the QwtText classes. The MathML renderer of
    the Qt4 solutions package is embedded. 
    work for all expressions/situations.
4.  New classes for navigating: QwtPanner, QwtMaginfier
5.  Spectrogram/Contour plots  and other classes for displaying
    raster data added.

Changes
-------
5.0.0 is by far the release with the most changes in the history of Qwt
- too many to make list.


Release 4.2.0/0.4.2
===================

License
--------
A couple of exceptions to the LGPL with the intention to allow static
linking with commercial applications. See COPYING.

Key features: 
-------------
1.  Designer plugin 
2.  Rich Text support ( f.e. E = m * c<sup>2</sup> ) added.
3.  QwtDial class family added (QwtDial, QwtCompass, QwtAnalogClock, ...)
4.  QwtPicker class family added. Includes QwtPlotZoomer, a complete
    implementation of recursive zooming.
5.  Device metrics independent printing of QwtPlot. (QPrinter::HighResolution)
6.  QwtPlot::setCurveBrush(), QwtCurve::setBrush() added. The area
    between curve and baseline will be filled with this brush.
7.  Rotation of axis tick labels added. Very useful for axis with long
    labels like time scales ...
8.  Added a new abstract QwtData class to plot data from almost any type of
    container class.  
9.  QwtDoublePoint, QwtDoubleSize, QwtDoubleRect double counterparts
    for QPoint, QSize, QRect.
10. First steps to support Qtopia. All examples can be compiled and started
    in the qvfb emulator.


Changes:
---------
1.  Rewrite of QwtLegend/QwtLegendItem (no QTable anymore)
2.  Each plot item will be painted, even if one of the axis it is 
    attached to is disabled. (like in all other releases beside 0.4.1)
3.  Code for double buffering moved to a new class QwtPaintBuffer.
    Double buffering can be enabled/disabled now. 
4.  QwtPainter, QwtMetricsMap, QwtLayoutMetrics added
    Hide paint device metrics dependencies.
5.  Layout code rewritten and moved to a new class QwtPlotLayout
    New layout options canvasMargin(), alignCanvasToScales()
6.  QwtPlot: sizeHint() != minimumSizeHint()
9.  Internal plot data are private again. A couple of get methods 
    added instead.
10. canvas repaints triggered by paint events. Enables event filtering
11. QwtPlot::drawCanvasItems added. In opposite to QwtPlot::drawCanvas
    it is used by the printing code too.
12. qwtMax, qwtMin, qwtInt mapped to QMAX, QMIN, qRound from qglobal.h
13. operator= for plot item classes changed.
14. readOnly property added for sliders.
15. valid flag added for QwtDblRange
16. QwtCounter wrap around policy: a counter under- or overflow sets
    focus to the smallest up/down button and disables counting.
    A space bar keypress release event re-enables counting.
17. QwtPushButton added. A class that adds rich text and alignments
    features to QPushButton, like they are used in QLabel
18. Clipped painting code moved from QwtCurve to QwtPainter/QwtRect
19. Canvas cache added to optimize trivial repaints.
20. QwtPlot::drawCurve added for incremental curve data
21. QwtSliderBase, readOnly, isValid added
22. Added filtering of the colors of the title and scales to QwtPrintFilter.
23. Support of QT_NO_CAST_ASII and QT_NO_COMPAT added

libsmi/tools/smidump.c
/*
 * smidump.c --
 *
 *      Dump a MIB module conforming to a given format.
 *
 * Copyright (c) 1999 Frank Strauss, Technical University of Braunschweig.
 * Copyright (c) 1999 J. Schoenwaelder, Technical University of Braunschweig.
 *
 * See the file "COPYING" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Id: smidump.c 7870 2008-03-11 19:29:58Z schoenw $
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_WIN_H
#include "win.h"
#endif

#include "smi.h"
#include "shhopt.h"
#include "smidump.h"



static void help(void);
static void usage(void);
static void version(void);
static void config(char *filename);
static void level(int lev);
static void quiet(void);
static void preload(char *module);
static void unified(void);
static void format(char *form);


static int flags;
static int mFlag = 0;	/* show the name for error messages */
static int sFlag = 0;	/* show the severity for error messages */
static int kFlag = 0;	/* keep going after severe errors */
static SmidumpDriver *driver;
static SmidumpDriver *firstDriver;
static SmidumpDriver *lastDriver;
static SmidumpDriver *defaultDriver;
static char *output;

static int opts;
static optStruct *opt;
static optStruct genericOpt[] = {
    /* short long              type        var/func       special       */
    { 'h', "help",           OPT_FLAG,   help,          OPT_CALLFUNC },
    { 'V', "version",        OPT_FLAG,   version,       OPT_CALLFUNC },
    { 'c', "config",         OPT_STRING, config,        OPT_CALLFUNC },
    { 'l', "level",          OPT_INT,    level,         OPT_CALLFUNC },
    { 'p', "preload",        OPT_STRING, preload,       OPT_CALLFUNC },
    { 'l', "level",          OPT_INT,    level,         OPT_CALLFUNC },
    { 'q', "quiet",          OPT_FLAG,   quiet,         OPT_CALLFUNC },
    { 'm', "error-names",    OPT_FLAG,   &mFlag,        0 },
    { 's', "severity",       OPT_FLAG,   &sFlag,        0 },
    { 'u', "unified",        OPT_FLAG,   unified,       OPT_CALLFUNC },
    { 'f', "format",         OPT_STRING, format,        OPT_CALLFUNC },
    { 'o', "output",         OPT_STRING, &output,       0            },
    { 'k', "keep-going",     OPT_FLAG,	 &kFlag,	0 },
    { 0, 0, OPT_END, 0, 0 }  /* no more options */
};



void *xmalloc(size_t size)
{
    char *m = malloc(size);
    if (! m) {
	fprintf(stderr, "smidump: malloc failed - running out of memory\n");
	exit(1);
    }

    return m;
}



void *xrealloc(void *ptr, size_t size)
{
    char *m = realloc(ptr, size);
    if (! m) {
	fprintf(stderr, "smidump: realloc failed - running out of memory\n");
	exit(1);
    }
    return m;
}



void *xcalloc(size_t nmemb, size_t size)
{
    char *m = calloc(nmemb, size);
    if (! m) {
	fprintf(stderr, "smidump: calloc failed - running out of memory\n");
	exit(1);
    }
    return m;
}



char *xstrdup(const char *s)
{
    char *m = strdup(s);
    if (! m) {
	fprintf(stderr, "smidump: strdup failed - running out of memory\n");
	exit(1);
    }
    return m;
}



void xfree(void *ptr)
{
    free(ptr);
}



void smidumpRegisterDriver(SmidumpDriver *driver)
{
    int i;
    
    if (!firstDriver) {
	firstDriver = driver;
	lastDriver = driver;
    } else {
	lastDriver->next = driver;
	lastDriver = driver;
    }

    for (i = 0; driver->opt && driver->opt[i].type != OPT_END; i++) {
	opt = xrealloc(opt, (opts+1) * sizeof(optStruct));
	memcpy(&opt[opts], &opt[opts-1], sizeof(optStruct));
	opt[opts-1].shortName = 0;
	opt[opts-1].longName  = xmalloc(strlen(driver->name) +
				 strlen(driver->opt[i].name) + 2);
	sprintf(opt[opts-1].longName, "%s-%s",
		driver->name, driver->opt[i].name);
	opt[opts-1].type      = driver->opt[i].type;
	opt[opts-1].arg       = driver->opt[i].arg;
	opt[opts-1].flags     = driver->opt[i].flags;
	opts++;
    }
}



static void formats()
{
    SmidumpDriver *driver = firstDriver;
    
    for (driver = firstDriver; driver; driver = driver->next) {
	fprintf(stderr, "  %-14s: %s\n", driver->name,
 		driver->descr ? driver->descr : "...");
    }
}



static void usage()
{
    int i;
    SmidumpDriver *driver;
    char *value = NULL;
#ifdef _MSC_VER
    #if _MSC_VER >= 1400
/*
   %n in printf is a security vulnerability. Ref:
   http://en.wikipedia.org/wiki/Format_string_vulnerabilities
   MS decided it was important enough to disble it by default. Ref:
   "ms-help://MS.VSCC.v80/MS.MSDN.v80/MS.VisualStudio.v80.en/dv_vccrt/html/77a854ae-5b48-4865-89f4-f2dc5cf80f52.htm
   Calling _set_printf_count_output() stops an invalid parameter crash.
*/
    int printf_state;
    #endif
#endif
    
    fprintf(stderr,
	    "Usage: smidump [options] [module or path ...]\n"
	    "  -V, --version        show version and license information\n"
	    "  -h, --help           show usage information\n"
	    "  -q, --quiet          do not generate any comments\n"
	    "  -c, --config=file    load a specific configuration file\n"
	    "  -p, --preload=module preload <module>\n"
	    "  -l, --level=level    set maximum level of errors and warnings\n"
	    "  -m, --error-names    print the name of errors in braces\n"
	    "  -s, --severity       print the severity of errors in brackets\n"
	    "  -f, --format=format  use <format> when dumping (default %s)\n"
	    "  -o, --output=name    use <name> when creating names for output files\n"
	    "  -u, --unified        print a single unified output of all modules\n"
	    "  -k, --keep-going     continue after serious parse errors\n\n",
	    defaultDriver ? defaultDriver->name : "none");

    fprintf(stderr, "Supported formats are:\n");
    formats();

    for (driver = firstDriver; driver; driver = driver->next) {
	if (! driver->opt) continue;
	fprintf(stderr, "\nSpecific option for the \"%s\" format:\n",
		driver->name);
	for (i = 0; driver->opt && driver->opt[i].type != OPT_END; i++) {
	    int n;
	    switch (driver->opt[i].type) {
	    case OPT_END:
	    case OPT_FLAG:
		value = NULL;
		break;
	    case OPT_STRING:
		value = "string";
		break;
	    case OPT_INT:
	    case OPT_UINT:
	    case OPT_LONG:
	    case OPT_ULONG:
		value = "number";
		break;
	    }
#ifdef _MSC_VER
    #if _MSC_VER >= 1400
	    printf_state=_set_printf_count_output(1);
    #endif
#endif
	    fprintf(stderr, "  --%s-%s%s%s%n",
		    driver->name, driver->opt[i].name,
		    value ? "=" : "",
		    value ? value : "",
		    &n);
#ifdef _MSC_VER
    #if _MSC_VER >= 1400
	    (void)_set_printf_count_output(printf_state);
    #endif
#endif
	    fprintf(stderr, "%*s%s\n",
		    30-n, "",
		    driver->opt[i].descr ? driver->opt[i].descr : "...");
	}
    }
}



static void help() { usage(); exit(0); }
static void version() { printf("smidump " SMI_VERSION_STRING "\n"); exit(0); }
static void config(char *filename) { smiReadConfig(filename, "smidump"); }
static void level(int lev) { smiSetErrorLevel(lev); }
static void quiet() { flags |= SMIDUMP_FLAG_SILENT; }
static void preload(char *module) { smiLoadModule(module); }
static void unified() { flags |= SMIDUMP_FLAG_UNITE; }

static void format(char *form)
{
    for (driver = firstDriver; driver; driver = driver->next) {
	if (strcasecmp(driver->name, form) == 0) {
	    break;
	}
    }
    if (!driver) {
	fprintf(stderr, "smidump: invalid dump format `%s'"
		" - supported formats are:\n", form);
	formats();
	exit(1);
    }
}



static void
errorHandler(char *path, int line, int severity, char *msg, char *tag)
{
    if (path) {
	fprintf(stderr, "%s:%d: ", path, line);
    }
    if (sFlag) {
	fprintf(stderr, "[%d] ", severity);
    }
    if (mFlag) {
	fprintf(stderr, "{%s} ", tag);
    }
    switch (severity) {
    case 4:
    case 5:
	fprintf(stderr, "warning: ");
	break;
    case 6:	
	fprintf(stderr, "info: ");
	break;
    }
    fprintf(stderr, "%s\n", msg);

    if (severity <= 0) {
	exit(1);
    }
}



int main(int argc, char *argv[])
{
    char *modulename;
    SmiModule *smiModule;
    int smiflags, i;
    SmiModule **modv = NULL;
    int modc = 0;

    output = NULL;
    firstDriver = lastDriver = defaultDriver = NULL;

    opts = sizeof(genericOpt) / sizeof(optStruct);
    opt = xmalloc(sizeof(genericOpt));
    memcpy(opt, genericOpt, sizeof(genericOpt));

    initCm();
    initCorba();
#if 0
    initFig();
#endif
    initIdentifiers();
    initImports();
    initJax();
    initMetrics();
    initMosy();
    initNetsnmp();
    initPerl();
    initPython();
    initSming();
    initSmi();                defaultDriver = lastDriver;
    initSppi();
#if 0
    initSql();
#endif
    initScli();
    initSvg();
    initTree();
    initTypes();
    initSizes();
    initXml();
    initXsd();
    initCompliances();
    initYang();
    initBoilerplate();
    
    for (i = 1; i < argc; i++)
	if ((strstr(argv[i], "-c") == argv[i]) ||
	    (strstr(argv[i], "--config") == argv[i])) break;
    if (i == argc) 
	smiInit("smidump");
    else
	smiInit(NULL);

    flags = 0;
    driver = defaultDriver;

    optParseOptions(&argc, argv, opt, 0);

    if (!driver) {
	fprintf(stderr, "smidump: no dump formats registered\n");
	smiExit();
	exit(1);
    }
    
    if (sFlag || mFlag) {
	smiSetErrorHandler(errorHandler);
    }

    smiflags = smiGetFlags();
    smiflags |= SMI_FLAG_ERRORS;
    smiflags |= driver->smiflags;
    smiSetFlags(smiflags);

    if (flags & SMIDUMP_FLAG_UNITE && driver->ignflags & SMIDUMP_DRIVER_CANT_UNITE) {
	fprintf(stderr, "smidump: %s format does not support united output:"
		" ignoring -u\n", driver->name);
	flags = (flags & ~SMIDUMP_FLAG_UNITE);
    }

    if (output && driver->ignflags & SMIDUMP_DRIVER_CANT_OUTPUT) {
	fprintf(stderr, "smidump: %s format does not support output option:"
		" ignoring -o %s\n", driver->name, output);
	output = NULL;
    }

    modv = (SmiModule **) xmalloc((argc) * sizeof(SmiModule *));
    modc = 0;
    
    for (i = 1; i < argc; i++) {
	modulename = smiLoadModule(argv[i]);
	smiModule = modulename ? smiGetModule(modulename) : NULL;
	if (smiModule) {
	    if ((smiModule->conformance) && (smiModule->conformance < 3)) {
		flags |= SMIDUMP_FLAG_ERROR;
		if (! (flags & SMIDUMP_FLAG_SILENT)) {
		    fprintf(stderr,
			    "smidump: module `%s' contains errors, "
			    "expect flawed output\n",
			    argv[i]);
		}
	    }
	    modv[modc++] = smiModule;
	} else {
	    fprintf(stderr, "smidump: cannot locate module `%s'\n",
		    argv[i]);
	}
    }

    if (! (flags & SMIDUMP_FLAG_ERROR) || kFlag) {
	(driver->func)(modc, modv, flags, output);
    } else {
	if (! (flags & SMIDUMP_FLAG_SILENT)) {
	    fprintf(stderr,
		    "smidump: aborting due to severe parsing errors\n"
		    "smidump: use the -k option to force continuation\n");
	}
    }

    smiExit();

    if (modv) xfree(modv);
    
    return 0;
}
